gems/pending/disk/modules/VhdxDisk.rb

```diff
@@ -0,0 +1,634 @@
+    'I<',  'block_size', # size of payload block in bytes between 1MB & 256MB
+    'b32', 'bit_fields',
+    # 'B1',  'leave_blocks_allocated',
+    # 'B1',  'has_parent',
+    # 'B30', 'reserved',
+  ])
+  SIZEOF_VHDX_FILE_PARAMETERS        = VHDX_FILE_PARAMETERS.size
+
+  VHDX_VIRTUAL_DISK_SIZE             = BinaryStruct.new([
+    'Q<', 'virtual_disk_size' # size in bytes.  must be a multiple of logical sector size
+  ])
+  SIZEOF_VHDX_VIRTUAL_DISK_SIZE = VHDX_VIRTUAL_DISK_SIZE.size
+
+  VHDX_LOGICAL_SECTOR_SIZE = BinaryStruct.new([
+    'I<', 'logical_sector_size' # size in bytes.  must be 512 or 4096
+  ])
+  SIZEOF_VHDX_LOGICAL_SECTOR_SIZE = VHDX_LOGICAL_SECTOR_SIZE.size
+
+  VHDX_PHYSICAL_SECTOR_SIZE = BinaryStruct.new([
+    'I<', 'physical_sector_size' # size in bytes.  must be 512 or 4096
+  ])
+  SIZEOF_VHDX_PHYSICAL_SECTOR_SIZE   = VHDX_PHYSICAL_SECTOR_SIZE.size
+
+  VHDX_PAGE_83_DATA                  = BinaryStruct.new([
+    'A16', 'page_83_data' # unique guid
+  ])
+  SIZEOF_VHDX_PAGE_83_DATA           = VHDX_PAGE_83_DATA.size
+
+  VHDX_PARENT_LOCATOR_HEADER         = BinaryStruct.new([
+    'A16', 'locator_type',      # guid with type of the parent virtual disk
+    'S<',   'reserved',
+    'S<',   'key_value_count',  # number of key-value pairs for the parent locator
+  ])
+  SIZEOF_VHDX_PARENT_LOCATOR_HEADER  = VHDX_PARENT_LOCATOR_HEADER.size
+  VHDX_PARENT_LOCATOR_TYPE_GUID      = 0xB04AEFB7D19E4A81B78925B8E9445913
+
+  VHDX_PARENT_LOCATOR_ENTRY          = BinaryStruct.new([
+    'I<',   'key_offset',       # offset within the metadata item of key
+    'I<',   'value_offset',     # offset within the metadata item of value
+    'S<',   'key_length',       # length of the entry's key
+    'S<',   'value_length',     # length of the entry's value
+  ])
+  SIZEOF_VHDX_PARENT_LOCATOR_ENTRY = VHDX_PARENT_LOCATOR_ENTRY.size
+
+  attr_reader :file_identifier_signature, :vhdx_header_signature, :dInfo
+  def d_init
+    @diskType             = "Vhdx"
+    @blockSize            = 0
+    @virtual_disk_size    = 0
+    @logical_sector_size  = 0
+    @physical_sector_size = 0
+    @payload_block_size   = 0
+    @has_parent           = nil
+    @parent_locator       = nil
+    @file_name            = dInfo.fileName
+    @hyperv_connection    = nil
+    @vhdx_file            = connection_to_file(dInfo)
+    @converter            = Encoding::Converter.new("UTF-16LE", "UTF-8")
+    header_section
+  end
+
+  def connection_to_file(dInfo)
+    if dInfo.mountMode.nil? || dInfo.mountMode == "r"
+      dInfo.mountMode     = "r"
+      file_mode           = "r"
+    elsif dInfo.mountMode == "rw"
+      file_mode           = "r+"
+    else
+      raise "Unrecognized mountMode: #{dInfo.mountMode}"
+    end
+    if dInfo.hyperv_connection
+      @vhdx_file = connect_to_hyperv(dInfo.hyperv_connection)
+    else
+      @vhdx_file = MiqLargeFile.open(@file_name, file_mode)
+    end
+    @vhdx_file
+  end
+
+  def d_read(pos, len)
```

Comment:
I wonder if we could run VCR like tests of this method by running a full end to end test of some scvmm disks, recording the inputs and outputs as test cases.  It might help better understand the conditionals in this method.